src/uffd_handler/src/prefaulter.rs

```diff
@@ -45,11 +46,108 @@ impl UffdPrefaulter {
+            .max(region.mapping.base_host_virt_addr);
+
+        // Page faults must be handled in units of pages. `page_fault_start_addr` already represents
+        // the start of a page and `bytes_before` is guaranteed to be a multiple of page size.
+        // However, the region start might not be, so the address must be rounded down to page
+        // boundary.
+        prefault_start_addr = page_start_of_addr(prefault_start_addr)?;
+
+        // If the start of the memory region has already been brought up into RAM
+        // (if for example a page fault was served by copying in RAM more than the faulting page)
+        // UFFD_COPY will return `EEXIST`. We need to find the first page not brought into memory
+        // after `prefault_start_addr`.
+        // This unwrap is safe because it is guaranteed that the key is the start of a page
+        // within the current region.
+        if *region.page_states.get(&prefault_start_addr).unwrap() == MemPageState::FromFile {
+            // Look for an uninitialized page up to the faulting page, which we know for
+            // sure that is not present in memory.
+            while prefault_start_addr < page_fault_start_addr {
+                prefault_start_addr += page_size;
+                // Found the first Uninitialized occurrence, so prefaulting will start from there.
+                // This unwrap is safe because it is guaranteed that the key is the start of a page
+                // within the current region.
+                if *region.page_states.get(&prefault_start_addr).unwrap()
+                    == MemPageState::Uninitialized
+                {
+                    break;
+                }
+            }
+        }
+
+        Ok(prefault_start_addr)
+    }
+
+    /// Return the end address of the chunk to prefault into memory. The address is obtained
+    /// by adding `bytes_after` amount to the end of the faulting page. If the end address
+    /// exceeds the current region, the chunk is truncated to fit into the memory region
+    /// (the end of the current memory region is returned).
+    /// When there is no customized prefaulting configured, the minimum chunk to copy to RAM
+    /// is the faulting page.
+    pub fn get_prefaulting_end_address(
+        &self,
+        page_fault_start_addr: usize,
+        region: &MemRegion,
+    ) -> Result<usize, HandlerError> {
+        let page_size = get_page_size().map_err(HandlerError::PageSize)?;
+
+        // This will not overflow because it has been verified beforehand during mappings
+        // deserialization.
+        let region_end = region.mapping.base_host_virt_addr + region.mapping.size;
+        // We add `page_size` because we always want to at least bring into RAM the faulting page,
+        // even if customized prefaulting is not configured. If the result exceeds the current
+        // region, the end of the region is returned instead.
+        Ok(self
+            .bytes_after
```

Comment:
What makes it impossible for some pages being already prefaulted between the currently faulting page and `bytes_after` so we don't need a loop similar to the one in `get_prefaulting_start_address `?

Reply:
The userfaultfd already takes care of this and stops copying when it encounters a page that has already been brought into memory. It doesn't error out, it just returns the number of bytes that it managed to bring into RAM. It is redundant for us to perform the same checks and verify each page specifically, as long as the faulting page is served.